lib/OpenQA/WebAPI/Plugin/ObsRsync/Runner.pm

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2019 SUSE Linux GmbH
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package OpenQA::WebAPI::Plugin::ObsRsync::Runner;
+use strict;
+use warnings;
+use IPC::System::Simple qw(system $EXITVAL);
+
+sub newRunner {
+    my $app = shift;
+    $app->minion->add_task(
+        obs_rsync => sub {
+            my ($job, $args) = @_;
+            eval { system([0], "bash", @$args); 1 };
+            $app->minion->unlock('obs_rsync_lock');
+            $job->finish($EXITVAL);
+        });
+}
+
+my $lock_timeout = 3600;
+
+sub Run {
+    my ($app, $home, $limit, $retry_timeout, $folder) = @_;
+    my $minion = $app->minion;
+    my @args   = ($home . "/rsync.sh", $folder);
+
+    my $bool = $minion->lock('obs_rsync_lock', $lock_timeout, {limit => $limit});
+    if (!$bool) {
+        return 1 unless $retry_timeout;
+        sleep $retry_timeout;
+        $minion->lock('obs_rsync_lock', $lock_timeout, {limit => $limit}) or return 1;
+    }
+    my $id = $minion->enqueue(obs_rsync => [\@args]);
+    return 1 if not $id;
+    $minion->job($id)->start();
```

Comment:
Don't do this. A gru worker should process the job.

Reply:
Did you mean "the gru worker"? Or how many gru workers a typical environment has? If it is only one - then it will not be able to cope with the load, which is required to sync repos from OBS for dozen of projects. And in general - this plugin's job shouldn't be delayed waiting in a queue, (so increasing number of workers probably is not the answer as well). It must be performed soon after the request was received; or an error returned if some limit is reached (or maybe retried with a brief delay).

Reply:
But you block the HTTP process and then *nothing* will handle *anything*

Reply:
There aren't that many projects to sync at a given time - and our old solution synced only max 2 at a time. So having 2 GRU workers will do

Reply:
Gru/Minion is a daemon process that forks up to X (think the default is 4 for gru, but it can be reconfigured to a higher value) processes on-demand to run jobs in parallel.

Reply:
Old solution did sync projects sequentially in few batches, so it is not the same. I don't mind using GRU as long as everyone agrees that following are not problems:
* other regular GRU tasks (like saving needles or download assets) can be delayed for 10-20 minutes, maybe more (assuming my understanding is correct);
* occasional 20-40 min delay before rsync starts will not only delay testing outcome, but also will lead to a chance that binaries will be overwritten by next OBS build publishing: that will result in test failures, which are tricky to troubleshoot. (And old approach did have smaller chance to hit such problem because it did check project status using api request).

Reply:
@andrii-suse If you're worried about delaying other gru jobs you can just use a lock to limit sync jobs to say 3 at a time and make sure gru is configured to have a pool of 6 (values chosen arbitrarily).

Reply:
I'm afraid we need to take care that there's always a "free slot" for saving needles.

Reply:
That's what minion workers have queues for. Create one for downloads

Reply:
@kraih delaying sync jobs will directly lead to the situation explicitly mentioned in previous comment. The Plugin's requests must be served immediately with high priority and all what you suggest is completely opposite to what is needed. I have no problem to repeat that delaying request for 20 min may lead to bad consequences, which are hard to troubleshoot. It is better to just reject request than delay it (At least I don't want to ship software which can lead to such situations). But rejecting at small limit is counter productive as well.
It is possible to implement some workarounds to avoid such bad consequences, but I don't see it as any good solution.
In other  words job queue only hurts and doesn't help here. Especially job queue shared with other tasks. All is needed is simple forking with sane limit of active forks at say 10 (just to have some limit), and reject requests if limit is reached. In reality 10 concurrent requests will probably never happen. Maybe 5 concurrent requests will never happen, but we still want to have a limit to honor Moore's law.
At the same time I agree that configuring gru forks at max 10 should be practically enough, but I don't want to take responsibility for situations when it is not, because those situations will be painful.

Reply:
There are many solutions available, depending on what problems we actually encounter. We have tools such as multiple named queues for routing jobs to different worker daemons (like one dedicated to downloading), locks to enforce concurrency limits (unique or fixed size pool), priorities to make sure needle handling jobs are always first, and the number of jobs each worker handles in parallel can even be changed on the fly with remote control commands.

Just use Minion/Gru the way it's supposed to be used and it will be fine. 😉

Reply:
I don't see where the requirement come from that you bring up against a job queue.

Reply:
Ok the requirement was to get rsync as fast as possible, because any delay will complicate handling and may result in inconsistent result. But since physical overload is concern - gru queue was introduced with best effort to check project status in Obs before starting rsync